### ISSUE TITLE
mAP50 is not correct since it use probIou.

### ISSUE THREAD

**Issue by dansontong** [ultralytics/ultralytics#15753]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

to Validate the performance of model, we need to compute the real IOU between prediction and ground  truth, but yolov8-obb use the probIou for computing mAP which is not correct.
the key code is in ultralytics/models/yolo/obb/val.py, _process_batch(), iou = **batch_probiou**(..)

### Environment

no need to provide

### Minimal Reproducible Example

# ultralytics/models/yolo/obb/val.py, _process_batch()
iou = batch_probiou(gt_bboxes, torch.cat([detections[:, :4], detections[:, -1:]], dim=-1)) 

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@dansontong thank you for pointing this out. We recommend updating to the latest version of the package to see if the issue persists. If it does, please open a new issue with detailed information so we can investigate further.

**Comment by dansontong** [comment]
@glenn-jocher I find this problem is latest version in main branch.  when yolov8-obb caculate true-positive(tp) in first line of _process_batch() in ultralytics/models/yolo/obb/val.py, it use 'iou = batch_probiou()', I think this is detailed enough.

by the way, the function compute_ap(recall, precision) in ultralytics/utils/metrics.py, line 553, 'mpre = np.concatenate(([1.0], precision, [0.0])) ' I think it's not right, since the precision[-1](corresponding to recall=1) not supposed to be 0. in many cases, precision can larger than 0 when recall=1.

**Comment by glenn-jocher** [comment]
@dansontong thank you for the detailed information. We appreciate your feedback. Please ensure you are using the latest version of the package. If the issue persists, we recommend opening a new issue with this detailed information so our team can investigate further.

**Comment by liu15509348793** [comment]
> 我发现这个问题是 main 分支中的最新版本。当 yolov8-obb 在 ultralytics/models/yolo/obb/val.py 中 _process_batch（） 的第一行计算真阳性（TP）时，它使用 'iou = batch_probiou（）'，我认为这已经足够详细了。
> 
> 顺便说一句，ultralytics/utils/metrics.py， line 553 中的函数 compute_ap（recall， precision） 'mpre = np.concatenate（（[1.0]， precision， [0.0]）） ' 我认为这是不对的，因为 precision[-1]（对应于 recall=1）不应该是 0。在许多情况下，当 recall=1 时，精度可以大于 0。

Hello, I also found the same problem, did you solve it?

**Comment by glenn-jocher** [comment]
Hello, this issue is under review. Please ensure you're using the latest version. If it persists, consider opening a new issue with detailed information for further investigation.

**Comment by liu15509348793** [comment]
> 您好，此问题正在审核中。请确保您使用的是最新版本。如果问题仍然存在，请考虑打开一个包含详细信息的新问题，以便进一步调查。

Thanks for your reply, will this issue be fixed in the near future? Just now another person gave me the following solution, I don't know if this works?
from ultralytics.utils.ops import xywhr2xyxyxyxy      
from ultralytics import YOLO
import ultralytics.models.yolo.obb.val as val
import ultralytics.utils.metrics as metrics
import torch
from shapely.geometry import Polygon

def rbox_iou(g, p):
    g = Polygon(g.cpu().numpy())
    p = Polygon(p.cpu().numpy())        
    g = g.buffer(0)
    p = p.buffer(0)
    if not g.is_valid or not p.is_valid:
        return 0
    inter = Polygon(g).intersection(Polygon(p)).area
    union = g.area + p.area - inter
    if union == 0:
        return 0
    else:
        return inter / union

def batch_probiou_monkey_patch(obb1, obb2, eps=1e-7):
  obb1 = xywhr2xyxyxyxy(obb1)
  obb2 = xywhr2xyxyxyxy(obb2)
  N, M = obb1.shape[0], obb2.shape[0]
  similarities = torch.zeros((N, M))

  for i, obb1_i in enumerate(obb1):
        for j, obb2_j in enumerate(obb2):
          similarities[i, j] = rbox_iou(obb1_i, obb2_j)
  return similarities.to("cuda:0")
val.batch_probiou = batch_probiou_monkey_patch
metrics.batch_probiou = batch_probiou_monkey_patch

model = YOLO("yolov8n-obb.pt")
results = model.val(data="dota8.yaml")
